### ISSUE TITLE
Fix internal self-referencing issues when manipulating data.table‘s

### ISSUE THREAD

**Issue by hongyuanjia** [SebKrantz/collapse#136]
Thanks for this amazing plugin. I did some very coarse benchmarking about {collapse} and {data.table} and got pretty impressed. Very looking forward to seeing the multi-threaded version as you mentioned by the end of this year. I am trying to replace some of my {data.table} workflow with {collapse}.

Currently `settransform()` did not work if {collapse} was not attached.

``` r
head(collapse::ftransform(airquality, Ozone = -Ozone))
#>   Ozone Solar.R Wind Temp Month Day
#> 1   -41     190  7.4   67     5   1
#> 2   -36     118  8.0   72     5   2
#> 3   -12     149 12.6   74     5   3
#> 4   -18     313 11.5   62     5   4
#> 5    NA      NA 14.3   56     5   5
#> 6   -28      NA 14.9   66     5   6
head(collapse::settransform(airquality, Ozone = -Ozone))
#> Error in ftransform(airquality, Ozone = -Ozone): could not find function "ftransform"
```

<sup>Created on 2021-03-13 by the [reprex package](https://reprex.tidyverse.org) (v1.0.0)</sup>

BTW,  to my understanding, `data.table::set*` functions are not drop-in replacements for `data.table::set*`. The latter modifies the passed data internally at the C level which will be memory efficient. But `collapse::settransform()` did not. In this sense, from a performance perspective, it makes no difference between `airquality <- collapse::ftransform(airquality, Ozone = -Ozone)` and `collapse::settransform(airquality, Ozone = -Ozone)` right? Please correct if I am wrong.

**Comment by SebKrantz** [comment]
Hi, thanks. So indeed what settransform does is evaluate the expression x <- ftransform(x, ...) in the parent environment. Since collapse was not attached ftransform is not available. It makes no performance difference so you can use x <- collapse::ftransform(x, ...). Talking about performance in general, apart from convenience I actually don’t see the point of data.table’s set functions doing things by reference at C-level. I presume people think that <- deep copies the data frame whenever you do traditional programming, but that is not the case. It creates a shallow copy of the data.frame and only deep copies the columns you are actually replacing. Maybe in some earlier version of R <- used to make deep copies, but that is no longer the case. Maybe I am wrong, but I’d be very interested if you can show me large data examples where data.table’s supposedly more efficient := function is more efficient that settransform. In my benchmarks so far this has not been the case and since <- is directly called from C I saw no reason of writing settransform in C. So if you want settransform and setrename are nothing sophisticated, just convenience functions to ease your coding. I also think bu the way working with shallow copies as <- does is much safer because if something breaks in your code the original object is unaffected.

**Comment by hongyuanjia** [comment]
@SebKrantz Thanks for the detailed explanation!

I am trying to use collapse in my package, so directly attaching collapse will not be a good decision. I will directly use `x <- collapse::ftransform(x, ...)`.

For the general performance, I did a fairly simple benchmark between `collapse::get_vars() <-` and `data.table::set()`. And you are right. They give the same speed!

``` r
N=1e7; K=100

DT1 <- data.table::data.table(
  id1 = sample(sprintf("id%03d",1:K), N, TRUE),      # large groups (char)
  id2 = sample(sprintf("id%03d",1:K), N, TRUE),      # large groups (char)
  id3 = sample(sprintf("id%010d",1:(N/K)), N, TRUE), # small groups (char)
  id4 = sample(K, N, TRUE),                          # large groups (int)
  id5 = sample(K, N, TRUE),                          # large groups (int)
  id6 = sample(N/K, N, TRUE),                        # small groups (int)
  v1 =  sample(5, N, TRUE),                          # int in range [1,5]
  v2 =  sample(5, N, TRUE),                          # int in range [1,5]
  v3 =  sample(round(runif(100,max=100),4), N, TRUE) # numeric e.g. 23.5749
)

pryr::object_size(DT1)
#> Registered S3 method overwritten by 'pryr':
#>   method      from
#>   print.bytes Rcpp
#> 526 MB

DT2 <- data.table::copy(DT1)

rbenchmark::benchmark(
    collapse = collapse::get_vars(DT1, "v1") <- sample(10, N, TRUE),
    data.table = data.table::set(DT2, NULL, "v1", sample(10, N, TRUE)),
    replications = 10
)
#>         test replications elapsed relative user.self sys.self user.child
#> 1   collapse           10    5.30    1.000      5.25     0.05         NA
#> 2 data.table           10    5.39    1.017      5.32     0.08         NA
#>   sys.child
#> 1        NA
#> 2        NA
```

<sup>Created on 2021-03-13 by the [reprex package](https://reprex.tidyverse.org) (v1.0.0)</sup>

However, `:=` really comes handy for operations like `DT[var1 == value1, var2 := ...]`, i.e. update certain column values by reference, based on `i`. I am wondering if there is a way to achieve this using collapse?

By the way, it seems that after subsetting using `collapse::ss()` or equivalent, the result data.table can no longer be used in `data.table::set()` and `[.data.table` with `:=`. Have to run `data.table::setalloccol()` to make it data.table-compatible again. If you think is cumbersome to fix, it will be good to at least mention this in the documentation to rebind the users.

``` r
mt <- data.table::as.data.table(mtcars)
data.table::set(mt[1L], NULL, "add", 1L)[]
#>    mpg cyl disp  hp drat   wt  qsec vs am gear carb add
#> 1:  21   6  160 110  3.9 2.62 16.46  0  1    4    4   1

mt_ss <- collapse::ss(mt, 1L)
data.table::set(mt_ss, NULL, "add", 1L)[]
#> Error in data.table::set(mt_ss, NULL, "add", 1L): It appears that at some earlier point, names of this data.table have been reassigned. Please ensure to use setnames() rather than names<- or colnames<-. Otherwise, please report to data.table issue tracker.

mt_ss <- collapse::ss(mt, 1L)
mt_ss[, a := 1L]
#> Warning in `[.data.table`(mt_ss, , `:=`(a, 1L)): Invalid .internal.selfref
#> detected and fixed by taking a (shallow) copy of the data.table so that :=
#> can add this new column by reference. At an earlier point, this data.table
#> has been copied by R (or was created manually using structure() or similar).
#> Avoid names<- and attr<- which in R currently (and oddly) may copy the whole
#> data.table. Use set* syntax instead to avoid copying: ?set, ?setnames and ?
#> setattr. If this message doesn't help, please report your use case to the
#> data.table issue tracker so the root cause can be fixed or this message
#> improved.

# have to run setalloccol
mt_ss <- collapse::ss(mt, 1L)
data.table::setalloccol(mt_ss)
#>    mpg cyl disp  hp drat   wt  qsec vs am gear carb
#> 1:  21   6  160 110  3.9 2.62 16.46  0  1    4    4
mt_ss[, a := 1L]

mt_ss <- collapse::ss(mt, 1L)
data.table::setalloccol(mt_ss)
#>    mpg cyl disp  hp drat   wt  qsec vs am gear carb
#> 1:  21   6  160 110  3.9 2.62 16.46  0  1    4    4
data.table::set(mt_ss, NULL, "add", 1L)[]
#>    mpg cyl disp  hp drat   wt  qsec vs am gear carb add
#> 1:  21   6  160 110  3.9 2.62 16.46  0  1    4    4   1
```

<sup>Created on 2021-03-13 by the [reprex package](https://reprex.tidyverse.org) (v1.0.0)</sup>

**Comment by SebKrantz** [comment]
Thanks a lot for the feedback and examples. Yes, I have to work on making the integration with data.table deeper to sort out the issues resulting from a data.table not just being a data.frame with an added “data.table” class. I will look into this over the coming weeks and see what I can do. Regarding conditional  replacing, you need to used a vectorized switch like data.table::fifelse inside ftansform. Alternatively you can try to use the replace function in base R e.g. fransform(data, new = replace(cond, val)), if val is the subset of a vector you would have to put val[!cond]. Fast vectorized switches are also provided in the very nice “kit” package. Actually the author of that package contributed fifelse and fcase to data.table. I don’t think at this point that I’ll add a vectorized switch to collapse. I see it’s mission in providing advanced statistical tools at superior execution speeds, not something super general.

**Comment by hongyuanjia** [comment]
Thanks for suggesting the "kit" package. I did not know it.

> I don’t think at this point that I’ll add a vectorized switch to collapse. I see it’s mission in providing advanced statistical tools at superior execution speeds, not something super general.

That totally makes sense. In terms of speed, you really have done an impressive job! collapse package brings hope to me for solving the bottleneck of my package. Thanks for the great work

**Comment by SebKrantz** [comment]
The issues flagged here are fixed in the new release 1.6.0. collapse now allocates extra columns when manipulating data.table's, allowing seamless workflows. You can also now call settransform without attaching collapse.
